eventdiscovery-sdk/src/main/java/com/schedjoules/eventdiscovery/framework/serialization/commons/NestedBundle.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 SchedJoules
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.schedjoules.eventdiscovery.framework.serialization.commons;
+
+import android.content.Intent;
+import android.os.Bundle;
+
+import com.schedjoules.eventdiscovery.framework.serialization.Keys;
+import com.schedjoules.eventdiscovery.framework.utils.factory.Factory;
+import com.schedjoules.eventdiscovery.framework.utils.optionals.AbstractCachingOptional;
+
+import org.dmfs.optional.NullSafe;
+import org.dmfs.optional.Optional;
+
+
+/**
+ * @author Gabor Keszthelyi
+ */
+public final class NestedBundle extends AbstractCachingOptional<Bundle>
+{
+    public NestedBundle(final Intent intent)
+    {
+        super(new Factory<Optional<Bundle>>()
+        {
+            @Override
+            public Optional<Bundle> create()
+            {
+                return new NullSafe<>(intent.getBundleExtra(Keys.NESTED_BUNDLE));
```

Comment:
If you say it like that you could call every method a "conditional factory", but that's not the point.

Try to see it like this. An `Obligatory` like this:
```java
new Obligatory(optional, default);
```
is the declarative equivalent of these two expressions:
```java
optional.isPresent()? optional : new Present<>(default);

new Present(optional.value(default));
```
Also, note that `Obligatory` is a decorator to an existing `Optional`, it's not a type nor a "standalone class". I agree that there wouldn't be any point in an `Obligatory<T>` interface, because we could just use `<T>` instead. 

This `Optional` design is different from Java 8's `Optional` which is rather procedural in that you often use it to describe an algorithm ("if … is not present then fall back to …") whereas our `Optional` (and this `Obligatory` in particular) can be used to describe the *result* in a declarative way ("the value of …, falling back to … if necessary").

My idea is that we should not have to use `isPresent()` very often when we use optionals. For instance, if I pass an `Optional<Bundle>` to an `Extra`, the `Extra` should be absent if either the `Bundle` is absent or if it doesn't contain the extra. I don't want to call `isPresent` on the optional Bundle before I create the extra.

```java
  Optional<X> y = new Extra(new NestedBundle(intent));
```

Also it would be great if the same would work with a "builder":

```java
  Intent newIntent = new IntentBuilder("action").withExtra(key, new Extra(new NestedBundle(oldIntent))).build().
```
In that case `newIntent` should only contain the extra if it was present in `oldIntent`.

Or to give you a more practical example. `BaseActivity` contains this code:
```java
Extra<Integer> themeExtra = new Extra<>(Keys.THEME, getIntent());
if (themeExtra.isPresent())
{
      new IntentParameter<>(Keys.THEME, new IntBox(themeExtra.value())).into(intent);
}
return intent;
```

I think it should look like this:
```java
return new IntentBuilder(intent).withExtra(Keys.THEME, new Extra<>(Keys.THEME, getIntent())).build();
```
Resulting in an intent that contains the theme extra if and only if it was present in the original intent (i.e. build an Intent with whatever the current Intent contains as the theme)

So maybe also consider to update the builder adapter interface below with a method like this
```
  // add an Optional extra only if present
  <T> BuilderAdapter withExtra(Key<T> key, Optional<T> value);
```
(names are subject to change).


So, to get back to the original issue: You don't have to apply `Obligatory` in `NestedBundle` but if you don't you should be more clear when you use it.
Your original code was 
```java
NestedBundle(getIntent()).value(null);
```
Which doesn't make any sense. Why would it default to `null`? Is it always expected to be present then just call `value()` (yes, we can't create a custom Bundle, that's a pitty, but even if we could, the interface would be unsuitable for this) if the `NestedBundle` is allowed to be absent, either use a proper fallback (if the method or constructor you were passing this to requires a value) or allow the constructor to take an `Optional` if absent values are valid parameters too.